qa: clarify and document one assumeutxo test case with malleated snapshot

darosior opened the pull request:
The `feature_assumeutxo.py` functional test checks various errors with malleated snapshots. Some of these cases are brittle or use confusing and undocumented values. Fix one of those by using a clear, documented and forward-compatible value.

I ran across those when working on an unrelated changeset which affected the snapshot. It took me a while to understand where the seemingly magic byte string was coming from, so i figured it was worth proposing this patch on its own for the sake of making the test more maintainable.

See commit messages for details.

darosior committed:
qa: use a clearer and documented amount error in malleated snapshot

In the assumeutxo functional tests, the final test case with alternated UTxO data tests the error
raised when deserializing a snapshot that contains a coin with an amount not in range (<0 or
>MAX_MONEY).

The current malleation uses an undocumented byte string and offset which makes it hard to maintain.
In addition, the undocumented offset is set surprisingly high (39 bytes is well into the
serialization of the amount which starts at offset 36). Similarly the value is surprisingly small,
presumably one was adjusted for the other. But there is no comment explaining how they were chosen,
why not in a clearer manner and what they are supposed to represent.

Instead replace this seemingly magic value with a clear one, MAX_MONEY + 1, serialize the whole
value for the amount field at the correct offset, and document the whole thing for the next person
around.

TheCharlatan commented:
The third paragraph in the first commit messages mentions commits that don't exist here. I'm guessing that is a reference to the original patchset you developed this with?

darosior commented:
Ugh yes it is indeed. Will fix.

theStack reviewed (commented):
Concept ACK, I agree that the snapshot modification tests are in a bit of a messy state currently.

yancyribbens commented on test/functional/feature_assumeutxo.py:
```diff
@@ -133,13 +133,20 @@ def expected_error(msg):
             # (content, offset, wrong_hash, custom_message)
             [b"\xff" * 32, 0, "7d52155c9a9fdc4525b637ef6170568e5dad6fabd0b1fdbb9432010b8453095b", None],  # wrong outpoint hash
-            [(2).to_bytes(1, "little"), 32, None, "Bad snapshot data after deserializing 1 coins."],  # wrong txid coins count
+            [(2).to_bytes(1, "little"), 32, None, "Bad snapshot data after deserializing 1 coins"],  # wrong txid coins count
```
nit `1 coin`?

rkrux replied:
This is an expected error from the underlying RPC call. Although using `coin` here in the assertion won't fail the functional test but I don't think this makes any meaningful difference.

yancyribbens replied:
Ah, I see thanks.  I guess the underlying call would be more clear to return `coin(s)` although that's besides the point.

janb84 commented:
tACK [f43b6ba](https://github.com/bitcoin/bitcoin/pull/31907/commits/f43b6ba27690442721623a9fa57b3c59e55d8397)

- Ran unit and functional tests related to assumeutxo (test/functional/feature_assumeutxo.py).
- Ran all the functional tests
- Validated the code change.

I also wrote a [review note](https://github.com/janb84/Bitcoin-Core_review-notes/blob/0f6634af718c2acedba9de74884cb149c2a68e2a/reviews/PR-%2331907.md) on the PR

darosior commented on test/functional/feature_assumeutxo.py:
```diff
@@ -139,7 +139,14 @@ def expected_error(msg):
             [b"\x84\x58", 34, None, "Bad snapshot data after deserializing 0 coins"],  # wrong coin case with height 364 and coinbase 0
-            [b"\xCA\xD2\x8F\x5A", 39, None, "Bad snapshot data after deserializing 0 coins - bad tx out value"],  # Amount exceeds MAX_MONEY
+            [
+                # VARINT(Compressed(MAX_MONEY + 1)) + VARINT(0)
+                b"\xa0\xc8\xad\xb1\xd1\x83\xff\x01" + b"\x00",
```
I'm on the fence. I agree it's nicer but also it would 10x the size of the change with harder to review code. Thanks for the commit but i'd rather keep it this way for now, unless/until more reviewers believe it's worth pulling in a full varint implementation.

rkrux replied:
The current implementation certainly makes it easier to parse but for someone new to this piece of code it still feels like magic value as the amount compression logic and the varint serialisation logic are quite custom. To eliminate the magical-ness completely, I would support pulling in the said routines in the test framework. Some of them are already present in the `utxo-to-sqlite.py` tool.

fjahr replied:
I'm in favor of adding varint and compression implementation here since it's a more meaningful improvement and the code is already there. Also not sure why we need to add `VARINT(0)`?

darosior replied:
> Also not sure why we need to add `VARINT(0)`?

You mean in the comment? I guess it doesn't matter either way but it's clearer this way...

If like @Prabhat1308 you ask why it's necessary to have an empty varint in the serialization itself, this is simply because otherwise you would get a serialization error instead of a `MoneyRange` error.

To give more details:
- this test is checking the error returned there https://github.com/bitcoin/bitcoin/blob/79bbb381a1fd13ad456fde736b3c195a791d4e58/src/validation.cpp#L5958
- which requires a coin was successfully deserialized there https://github.com/bitcoin/bitcoin/blob/79bbb381a1fd13ad456fde736b3c195a791d4e58/src/validation.cpp#L5951
- which in turn requires to successfully deserialize a compressed txout https://github.com/bitcoin/bitcoin/blob/79bbb381a1fd13ad456fde736b3c195a791d4e58/src/compressor.h#L112-L115
- therefore my serialization uses an amount of `MAX_MONEY + 1` (to test the bound) and an empty scriptpubkey (because we don't care). The latter is serialized as an empty VarInt.

darosior commented:
Updated the commit message to remove the reference to later modification i had in my branch.

janb84 commented:
reACK [dedc2d9](https://github.com/bitcoin/bitcoin/commit/dedc2d9cecccb866230c07112d00288237d12eaa)

rkrux commented on test/functional/feature_assumeutxo.py:
Comment about the commit message:
Curious to know the reason behind using `UTx0` instead of `UTXO`.

darosior replied:
>  Curious to know the reason behind using UTx0 instead of UTXO.

:shrug: doesn't matter, but as it is an acronym it makes more sense to me to capitalize the first letter of each word an not the others.

> Can this be made more understandable?

That's a pretty vague request.. I tried to be as detailed as possible in the commit message since the change is not obvious. Could you point to what you don't understand exactly?

> What I can gather is this seems to make the test less tight around the actually expected error (?) and I don't understand how that would make the test better.

Indeed this makes the test less tight because the test is currently unnecessarily specific. It's supposed to be testing that a wrong txid count will cause a deserialization error, but it's currently tightened to one of the many deserialization errors that can happen. The error that happens will depend on the snapshot data itself, which may change, therefore this check is brittle. This commit changes the check to match on two of the many deserialization errors that can happen. This makes it slightly less brittle.

> I don't see what's the problem with that.

The issue is that the test does not check what is documented but something stricter and dependent on the snapshot data itself which may vary.

I think i'll just drop this commit from this PR and keep it with my other PR.

darosior replied:
I've now dropped this commit from this PR. I'll keep it with my followup branch for now.

rkrux reviewed (commented):
Concept ACK dedc2d9cecccb866230c07112d00288237d12eaa

fanquake commented:
@fjahr you might want to have a look here?

Prabhat1308 commented on test/functional/feature_assumeutxo.py:
```diff
@@ -133,13 +133,20 @@ def expected_error(msg):
-            [b"\xCA\xD2\x8F\x5A", 39, None, "Bad snapshot data after deserializing 0 coins - bad tx out value"],  # Amount exceeds MAX_MONEY
+            [
+                # VARINT(Compressed(MAX_MONEY + 1)) + VARINT(0)
+                b"\xa0\xc8\xad\xb1\xd1\x83\xff\x01" + b"\x00",
```
as the above comment has already asked , I can't seem to understand why `VARINT(0)` was added considering we have already exceeded the max amount in the first amount written itself.

darosior replied:
See https://github.com/bitcoin/bitcoin/pull/31907/commits/dedc2d9cecccb866230c07112d00288237d12eaa#r1978352110

Prabhat1308 reviewed (commented):
tACK [`dedc2d9`](https://github.com/bitcoin/bitcoin/pull/31907/commits/dedc2d9cecccb866230c07112d00288237d12eaa)

- tested the amount with compression and varint encoding to match .
- On the case of  implementing varint serialization and compression logic , it would be a great addition to bring in the logic to test_framework (or in any other PR , no preference to bring it in this one).

theStack committed:
test: introduce VARINT (de)serialization routines

theStack committed:
test: introduce output amount (de)compression routines

darosior commented:
Since it was tiny, [pretty specifc to my branch](https://github.com/bitcoin/bitcoin/pull/31907#issuecomment-2669777720) and [confused some reviewers](https://github.com/bitcoin/bitcoin/pull/31907#discussion_r1966033115), i dropped the first commit of this PR.

Since several reviewers were in favor, i included in this PR @theStack's commits to introduce varint and compression serialization primitives to the functional test framework. I adapted the commit to use these primitives instead of the documented bytestring i was using before.

I have also reworked the commit message a bit. Please let me know if it's still too hard to understand.

janb84 commented:
re ACK [e5ff4e4](https://github.com/bitcoin/bitcoin/commit/e5ff4e416ecc8a51367022eb8a7a291f8cbc0c65)

theStack reviewed (approved):
ACK e5ff4e416ecc8a51367022eb8a7a291f8cbc0c65

fjahr commented on test/functional/feature_assumeutxo.py:
```diff
@@ -139,7 +144,14 @@ def expected_error(msg):
+            [
+                # compressed txout value + scriptpubkey
+                ser_varint(compress_amount(MAX_MONEY + 1)) + ser_varint(0),
+                # txid + coins per txid + vout + coin height
```
``` suggestion
                # txid + coins per txid + vout + coin code (height*2 + coinbase)
```

fjahr reviewed (commented):
Code review ACK e5ff4e416ecc8a51367022eb8a7a291f8cbc0c65

Would be fine for me to leave comments for when these files are retouched. The title and description of the PR should still be updated to reflect the new content though.

darosior commented:
Thanks for the reviews. Updated title and description to only mention one of the test cases since i dropped the other commit. I don't think it's worth addressing the nits unless people feel too strongly.

yuvicc reviewed (commented):
tACK e5ff4e416ecc8a51367022eb8a7a291f8cbc0c65

- Ran all functional and unit test for assume utxo
- Tested with different invalid amounts
- Verified the serialization

ryanofsky commented:
Merged since it was requested in IRC. Could be useful to follow up on comments from https://github.com/bitcoin/bitcoin/pull/31907#pullrequestreview-2661853159 though

TheCharlatan commented:
Post-merge ACK

pablomartin4btc reviewed (commented):
post-merge ACK e5ff4e416ecc8a51367022eb8a7a291f8cbc0c65

The updated case looks obviously much clearer now (3rd commit body explains the change very well too) and I'm in favour of the [mentioned](https://github.com/bitcoin/bitcoin/pull/31907#pullrequestreview-2661853159) follow-ups.
